tests/probes/yamlfilecontent/test_probes_yamlfilecontent.xml

```diff
@@ -0,0 +1,136 @@
+
+  <generator>
+    <oval:product_name>yamlfilecontent</oval:product_name>
+    <oval:product_version>1.0</oval:product_version>
+    <oval:schema_version>5.10</oval:schema_version>
+    <oval:timestamp>2020-02-13T00:00:00-00:00</oval:timestamp>
+  </generator>
+
+  <definitions>
+
+    <definition class="compliance" version="1" id="oval:0:def:1"> <!-- comment="true" -->
+      <metadata>
+        <title></title>
+        <description></description>
+      </metadata>
+      <criteria operator="AND">
+        <criterion comment="get_root_scalar" test_ref="oval:0:tst:1"/>
+        <criterion comment="get_nested_scalar" test_ref="oval:0:tst:2"/>
+        <criterion comment="filtered_get" test_ref="oval:0:tst:3"/>
+        <criterion comment="filtered_get_multiple" test_ref="oval:0:tst:4"/>
+      </criteria>
+    </definition>
+
+    <definition class="compliance" version="1" id="oval:0:def:2"> <!-- comment="error" -->
+      <metadata>
+        <title></title>
+        <description></description>
+      </metadata>
+      <criteria operator="AND">
+        <criterion comment="non_scalar_get" test_ref="oval:0:tst:5"/>
+      </criteria>
+    </definition>
+
+  </definitions>
+
+  <tests>
+
+    <ind-def:yamlfilecontent_test version="1" id="oval:0:tst:1" check="all" comment="true">
+      <ind-def:object object_ref="oval:0:obj:1"/>
+      <ind-def:state state_ref="oval:0:ste:1"/>
+    </ind-def:yamlfilecontent_test>
+
+    <ind-def:yamlfilecontent_test version="1" id="oval:0:tst:2" check="all" comment="true">
+      <ind-def:object object_ref="oval:0:obj:2"/>
+      <ind-def:state state_ref="oval:0:ste:2"/>
+    </ind-def:yamlfilecontent_test>
+
+    <ind-def:yamlfilecontent_test version="1" id="oval:0:tst:3" check="all" comment="true">
+      <ind-def:object object_ref="oval:0:obj:3"/>
+      <ind-def:state state_ref="oval:0:ste:3"/>
+    </ind-def:yamlfilecontent_test>
+
+    <ind-def:yamlfilecontent_test version="1" id="oval:0:tst:4" check="all" comment="true">
+      <ind-def:object object_ref="oval:0:obj:4"/>
+      <ind-def:state state_ref="oval:0:ste:4"/>
+    </ind-def:yamlfilecontent_test>
+
+    <ind-def:yamlfilecontent_test version="1" id="oval:0:tst:5" check="all" comment="error">
+      <ind-def:object object_ref="oval:0:obj:5"/>
+    </ind-def:yamlfilecontent_test>
+
+  </tests>
+
+  <objects>
+
+    <ind-def:yamlfilecontent_object version="1" id="oval:0:obj:1">
+      <ind-def:path>/tmp</ind-def:path>
+      <ind-def:filename>openshift-logging.yaml</ind-def:filename>
+      <ind-def:yamlpath>/kind</ind-def:yamlpath>
```

Comment:
Current YAML Path definition is here: https://github.com/evgenyz/libyaml-yamlpath-filter/wiki/YAML-Path-(OVAL-subset). It became even smaller; for now there are only *keys*, *indexes* and *slices* (for lists, not for maps). And only dot notation (to be more compatible with JSONPath).

You can get the idea of what it is capable of right now here: https://github.com/evgenyz/libyaml-yamlpath-filter/blob/afb1fec90a5eb3974d1115bab1d1bfff4956b91a/test-paths.c#L170.

Reply:
> for now there are only _keys_, _indexes_ and _slices_ (for lists, not for maps).

I'm not sure if array slicing will be that useful. Can we be sure of the order of the items in a list?

So hash key searching is not supported? :cry:
I think key searching will be the most useful feature.

Reply:
There is a case already for slices (which could be used as *any* filter for sequences), in the example we have we are looking for at least one record of certain type. The path is `.spec.pipelines[:].inputSource` and result would be `['logs.app', 'logs.infra', 'logs.audit']`. We are after logs.audit.

Reply:
Key searching is another thing. You can address anything by an exact key. But there is no any kind of fuzzy key search right now, until we prove it to be useful. I try to keep it as dump as possible to actually have all the logic in OVAL territory.

Reply:
> Can we be sure of the order of the items in a list?

Yes, it'd be the same as in the document.

Reply:
> > Can we be sure of the order of the items in a list?
> 
> Yes, it'd be the same as in the document.

I'm more concerned about the order in the yaml file.
Is it the case that a field we are looking for, will always be the first entry in list?

Reply:
> There is a case already for slices (which could be used as _any_ filter for sequences), in the example we have we are looking for at least one record of certain type. The path is `.spec.pipelines[:].inputSource` and result would be `['logs.app', 'logs.infra', 'logs.audit']`. We are after logs.audit.

In this example you want to know if any pipeline has `logs.audit` as inputSource.
Can there be an use case in which you want to retrieve a certain field from a pipeline that has `inputSource: logs.audit`?

Reply:
No, of course it won't. But when we are dealing with arrays, we, in general, have to check all the items for that reason. I understand that `[type='logs.app']` feature is attractive. But, in addition to another round of filter library development (this check can't be done in a single pass), we would face the need for some type coercion in the probe, which we are trying to avoid.

Lets just stick to what we have right now, and evolve with the content, should the need arise.

Reply:
> Can there be an use case in which you want to retrieve a certain field from a pipeline that has inputSource: logs.audit?

*There are more things in heaven and Earth, Horatio, 
Than are dreamt of in your philosophy.* 

There *might* be, for sure. Even the aforementioned case could be rewritten as `.spec.pipelines[inputSource=logs.audit]` and then just have to check if results count >1. But I want to be sure that workaround is ugly enough or impossible before starting to deal with all the problems related to [XXX=YYY].

Reply:
Playing around with `yaml-get`:

```bash
$ yaml-get -p '.spec.pipelines[:].inputSource' openshift-logging.yaml
CRITICAL:  : is not an integer array slice at ':' in 'spec.pipelines[:].inputSource'.                                                                                                                              
$ yaml-get -p '.spec.pipelines[0:2].inputSource' openshift-logging.yaml                                                                                         
logs.app                                                                                                 
logs.infra                                                                                               
$ yaml-get -p '.spec.pipelines[0:5].inputSource' openshift-logging.yaml                                                                                         
logs.app                                                                                                 
logs.infra                                                                                               
logs.audit                                          
$ yaml-get -p '.spec.pipelines.inputSource' openshift-logging.yaml                                                                                             
logs.app                                                                                                 
logs.infra                                                                                                                                                                                                         
logs.audit          
```

Is the [python style slicing syntax](https://github.com/evgenyz/libyaml-yamlpath-filter/wiki/YAML-Path-(OVAL-subset)#slices-arrays) specific to [libyaml-yamlpath-fitler](https://github.com/evgenyz/libyaml-yamlpath-filter)?
`yaml-get` doesn't like it.

@evgeni Would behavior of the last command work with the filter?

Reply:
My (and Matej's) beloved example: https://docs.ansible.com/ansible/latest/modules/file_module.html#parameter-mode.

From some Ansible file you want to collect paths of files with access mode is `0777`. What will you write in `[mode=XXXXXX?]`. What will you do in the probe when you get this XXXXX (as a char*, with the only notion about type that it was explicitly quoted or wasn't)?

Reply:
I am not the Evgeni you're looking for ;)

Reply:
> Lets just stick to what we have right now, and evolve with the content, should the need arise.

I understand, we need to start writing content as soon as possible then, to collect use cases, even when the probe is not ready.

> But I want to be sure that workaround is ugly enough or impossible before starting to deal with all the problems related to [XXX=YYY].

At the same time, one of the goals of the probe is not to have to write workarounds or complex OVAL contents.

Reply:
@evgeni Apologies, :open_mouth:

Reply:
No worries :)

Reply:
> At the same time, one of the goals of the probe is to avoid need for workarounds or complex OVAL contents.

Balance. That is what we usually looking for :)

>  I understand, we need to start writing content as soon as possible then, to collect use cases, even when the probe is not ready.

Exactly. So I'm trying to create something to give our OCP colleagues tools to start working and pinpointing weak spots.

> Is the python style slicing syntax specific to libyaml-yamlpath-fitler?
yaml-get doesn't like it.

Yes, it is. And it is how things are done in JSONPath. I don't like YAML Paths' version because you can't collect the whole list with it.

YAML Paths' `'.spec.pipelines.inputSource'` is too ambiguous. It completely skips the list in the middle, inputSource may as well be just a key in pipelines dictionary, which is not acceptable at all. It is so because YAML Path uses so called *arrays pass-through*, which is not good from a security-related tool point of view, IMHO.

Reply:
BTW, if you want to play with *our* (he-he) YAML Path, go ahead: https://github.com/OpenSCAP/yaml-filter. We even have a console tool now :)